// iota-auth/lib/impl/iotauth.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const iotaSeed = require("iota-seed-generator");
const IOTA = require("iota.lib.js");
const moment = require("moment");
class IotAuth {
    constructor(seed, duration = Infinity, node = 'https://nodes.iota.cafe') {
        this.iotaClient = new IOTA({
            provider: node,
        });
        if (seed) {
            this.receiveSeed = seed;
        }
        this.duration = duration;
    }
    isTransactionValid(validationCode) {
        return __awaiter(this, void 0, void 0, function* () {
            const receiveSeed = yield this.getSeed();
            const accountData = yield this.getAccountData(receiveSeed);
            const transferObj = accountData.transfers[accountData.transfers.length - 1];
            const transfer = transferObj[0];
            try {
                let code = this.iotaClient.utils.extractJson(transferObj);
                code = JSON.parse(code);
                const isValidAddress = yield this.isValidAddress(transfer.address, accountData.transfers.length - 1);
                const isValidTimestamp = this.isValidTimestamp(transfer.timestamp);
                return (((code && code.code === validationCode) || !validationCode) &&
                    isValidAddress &&
                    isValidTimestamp);
            }
            catch (e) {
                return false;
            }
        });
    }
    getSeed() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.receiveSeed) {
                this.receiveSeed = yield this.generateNewSeed();
            }
            return this.receiveSeed;
        });
    }
    generateValidationCode() {
        return __awaiter(this, void 0, void 0, function* () {
            const seed = yield iotaSeed();
            return seed.slice(0, 6);
        });
    }
    getNewAddress(seed, options = { index: 0, returnAll: true }) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise(resolve => {
                this.iotaClient.api.getNewAddress(seed, options, (empty, addresses, transactions) => {
                    if (addresses instanceof Array) {
                        resolve(addresses[Math.min(options.index, addresses.length - 1)]);
                    }
                    else {
                        resolve(addresses);
                    }
                });
            });
        });
    }
    isValidAddress(receiveAddress, index) {
        return __awaiter(this, void 0, void 0, function* () {
            const correctAddress = yield this.getNewAddress(this.receiveSeed, {
                index,
                returnAll: true,
            });
            return correctAddress === receiveAddress;
        });
    }
    isValidTimestamp(timestamp) {
        const transactionTime = moment(timestamp * 1000);
        const now = moment();
        const diff = now.diff(transactionTime, 'minutes');
        return diff <= this.duration;
    }
    generateNewSeed() {
        return __awaiter(this, void 0, void 0, function* () {
            const seed = yield iotaSeed();
            return seed;
        });
    }
    getAccountData(seed) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => {
                this.iotaClient.api.getAccountData(seed, (error, accountData) => {
                    resolve(accountData);
                });
            });
        });
    }
}
exports.IotAuth = IotAuth;
